### PR TITLE
feat(pipettes): add capacitive sensor support

### PR DESCRIPTION
**Pull request by Laura-Danielle**
## Overview

Apologies for the PR being all over the place. I tried to split up the commits, at least, to make it more readable.

Overall, this PR adds support for the fdc1004 capacitive sensor. There is an ability to set an offset -- depending on the environment the sensor is in. The idea here for the offset specifically is that you would be able to get the new offset by sending a `BaselineSensorRequest`, and then you would send over a `WriteToSensor` request to update the current sensor offset on the capacitive sensor itself.

## Changelog
- Added in an `offset` parameter to the `Read/Baseline` requests which will determine whether you are taking a sensor read or an offset read (this is used for both the pressure and capacitive sensor)
- Added the ability to poll from multiple registers
- Modified polling so that it sends write/read requests to the i2c bus for a given register
- Created abstract base classes for the two callback types used for the sensor tasks
- overloaded the operator function of the callback so that you can send a final can message once the firmware has taken an average read for polling
- Added a test for the capacitive sensor

## Review Requests
Please let me know if anything is confusing!

### CODE REVIEW

File: include/sensors/core/callback_types.hpp [hunk 1]
```diff
@@ -0,0 +1,23 @@
+#pragma once
+
+namespace sensor_callbacks {
+using MaxMessageBuffer = std::array<uint8_t, 5>;
+
+// NOLINTNEXTLINE(cppcoreguidelines-special-member-functions)
+class SingleRegisterCallback {
+  public:
+    virtual void operator()(const MaxMessageBuffer &buffer) = 0;
```
Comment (amitlissack) [inline include/sensors/core/callback_types.hpp:9]: One of the expectations of an interface like this is that someone who needs to create an implementation will know what the methods need to do. 

I don't know when `operator()()` is called verses the other `operator`. What should happen in these functions?

Comment (Laura-Danielle) [reply include/sensors/core/callback_types.hpp:9]: I'll add documentation

File: include/sensors/core/callback_types.hpp [hunk 2]
```diff
@@ -0,0 +1,23 @@
+#pragma once
+
+namespace sensor_callbacks {
+using MaxMessageBuffer = std::array<uint8_t, 5>;
+
+// NOLINTNEXTLINE(cppcoreguidelines-special-member-functions)
+class SingleRegisterCallback {
```
[the remark this answers is not in the document]
Comment (Laura-Danielle) [reply include/sensors/core/callback_types.hpp:7]: I chose to do a pure virtual function because I honestly didn't know how else to type this callback. The reason for having two overloaded operator functions is so that the i2c task can separately choose to send a CAN message back. This is most relevant for polling a specific i2c bus where you'll want to accumulate the data without actually sending a message back.

File: include/sensors/core/tasks/capacitive_sensor_callbacks.hpp [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+        can_client.send_can_message(can_ids::NodeId::host, message);
+    }
+
+    void reset() {
+        number_of_reads = 1;
+        measurement = 0;
+    }
+
+    void set_number_of_reads(uint16_t number_of_reads) {
+        number_of_reads = number_of_reads;
```
Comment (amitlissack) [inline include/sensors/core/tasks/capacitive_sensor_callbacks.hpp:54]: This is a weird line to look at. Can you please make it 
```this->number_of_reads = number_of_reads;```?

File: include/sensors/core/tasks/capacitive_sensor_callbacks.hpp [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+        can_client.send_can_message(can_ids::NodeId::host, message);
+    }
+
+    void reset() {
+        measurement = 0;
+        number_of_reads = 1;
+    }
+
+    void set_number_of_reads(uint16_t number_of_reads) {
+        number_of_reads = number_of_reads;
```
Comment (amitlissack) [inline include/sensors/core/tasks/capacitive_sensor_callbacks.hpp:100]: Same as above

File: include/sensors/core/tasks/capacitive_sensor_task.hpp [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+        THRESHOLD = m.threshold;
+        auto message = can_messages::SensorThresholdResponse{
+            {}, SensorType::capacitive, THRESHOLD};
+        can_client.send_can_message(can_ids::NodeId::host, message);
+    }
+
+    InternalCallback internal_callback{};
+    sensor_task_utils::BitMode mode = sensor_task_utils::BitMode::MSB;
+    // 8 pF
+    uint32_t THRESHOLD = 0x8;
```
Comment (amitlissack) [inline include/sensors/core/tasks/capacitive_sensor_task.hpp:107]: These three are in all caps. Are they constants?

File: include/pipettes/core/tasks/i2c_task.hpp [hunk 6]
```diff
@@ -33,18 +33,46 @@ class I2CMessageHandler {
     void visit(ReadFromI2C &m) {
         i2c_device.central_receive(m.buffer.data(), m.buffer.size(), m.address,
                                    TIMEOUT);
-        m.client_callback(m.buffer);
+        m.handle_buffer(m.buffer);
+        m.client_callback();
     }
 
-    void visit(PollReadFromI2C &m) {
+    void visit(SingleRegisterPollReadFromI2C &m) {
```
Comment (amitlissack) [inline include/pipettes/core/tasks/i2c_task.hpp:40]: It looks like processing `SingleRegisterPollReadFromI2C` and `MultiRegisterPollReadFromI2C`  messages will keep this task busy for the duration of the poll. 

Is there a concern that blocking all other I2C communication during the poll can be a problem?

Should the polling be happening from a different task?

**Review by amitlissack — COMMENTED** [review]
A serious question about polling within the i2c task. Otherwise rock solid.

File: include/sensors/core/tasks/capacitive_sensor_task.hpp (hunk 5 above)
Comment (Laura-Danielle) [reply include/sensors/core/tasks/capacitive_sensor_task.hpp:107]: Sort of..you can change the threshold to something else from the python side but it shouldn't be changed frequently.

File: include/pipettes/core/tasks/i2c_task.hpp (hunk 6 above)
Comment (Laura-Danielle) [reply include/pipettes/core/tasks/i2c_task.hpp:40]: I wanted it to block i2c communication because you can't read/write from two things at the same time on the same bus which is why I had it block in the i2c task.

File: include/sensors/core/tasks/capacitive_sensor_task.hpp (hunk 5 above)
Comment (amitlissack) [reply include/sensors/core/tasks/capacitive_sensor_task.hpp:107]: Ok. Can we make them lower case or `const`?

File: include/pipettes/core/i2c_writer.hpp [hunk 7]
```diff
@@ -38,32 +41,108 @@ class I2CWriter {
...
+     * manipulation required for the specific task using the i2c bus. Takes in
+     * two buffers.
+     * @param reg: the register to read from, if relevant.
+     *
+     * A poll function automatically initiates a read/write to the specified
+     * register on the i2c bus.
+     */
+    void read(uint16_t device_address, SingleBufferTypeDef handle_callback,
+              SendToCanFunctionTypeDef client_callback, uint8_t reg = 0x0) {
         // We want to copy the callback every time as opposed to passing a
```
Comment (sfoster1) [inline include/pipettes/core/i2c_writer.hpp:60]: can we remove this comment? it's not quite accurate (we're moving, not copying) and it also doesnt mention why we'd want to do this

File: include/pipettes/core/i2c_writer.hpp [hunk 8]
```diff
@@ -38,32 +41,108 @@ class I2CWriter {
...
+     * @param client_callback: the function that will write a CAN return message
+     * to the client queue.
+     * @param single_callback: the function that will handle any data
+     * manipulation required for the specific task using the i2c bus.
+     * @param reg: the register to read from, if relevant.
+     *
+     * A poll function automatically initiates a read/write to the specified
+     * register on the i2c bus.
+     */
+    void single_register_poll(uint16_t device_address, uint8_t number_reads,
```
Comment (sfoster1) [inline include/pipettes/core/i2c_writer.hpp:89]: let's give `number_reads` a bigger datatype just to get ahead of ourselves

File: include/pipettes/core/i2c_writer.hpp [hunk 9]
```diff
@@ -38,32 +41,108 @@ class I2CWriter {
...
+                              uint16_t delay,
+                              SendToCanFunctionTypeDef client_callback,
+                              SingleBufferTypeDef handle_callback,
+                              uint8_t reg = 0x0) {
         // We want to copy the callback every time as opposed to passing a
```
Comment (sfoster1) [inline include/pipettes/core/i2c_writer.hpp:94]: remove or update

File: include/pipettes/core/tasks/i2c_task.hpp [hunk 10]
```diff
@@ -33,18 +33,52 @@ class I2CMessageHandler {
...
         auto empty_array = m.buffer;
         for (int i = 0; i < m.polling; i++) {
+            i2c_device.central_transmit(m.buffer.data(), m.buffer.size(),
+                                        m.address, TIMEOUT);
             i2c_device.central_receive(m.buffer.data(), m.buffer.size(),
                                        m.address, TIMEOUT);
-            m.client_callback(m.buffer);
+            m.handle_buffer(m.buffer);
             m.buffer = empty_array;
             i2c_device.wait_during_poll(m.delay_ms);
```
Comment (sfoster1) [inline include/pipettes/core/tasks/i2c_task.hpp:52]: we also might want to delay this to another pr but this is not how you want to do fixed-frequency polling since it adds the time taken for the message to the wait. If you pass 100ms here, then you'll be reading every 100ms + transmit/receive time. Usually what you want to do is take a timestamp before transmission begins and then use something like [`vTaskDelayUntil`](https://www.freertos.org/vtaskdelayuntil.html) to delay until `timestamp+100ms`, which gives you more consistent timing (or use a timer)

File: include/sensors/core/callback_types.hpp [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+    {cb.handle_data(buffer, buffer)};
+    {cb.send_to_can()};
+};
+
+template <SingleRegister scb>
+struct SingleRegisterCallback {
+    SingleRegisterCallback(scb &callback) : callback{callback} {}
+    scb &callback;
+
+    void operator()(const MaxMessageBuffer &buffer) {
```
Comment (sfoster1) [inline include/sensors/core/callback_types.hpp:45]: Let's not do this thing with the `operator()()` overloads. It's a really bad idea to make function overloads do totally different things, it makes code unreadable. Function overloads are best used to do fundamentally the same thing from somewhat different data - maybe a level of indirection, maybe a cast. Having two function overloads that do two entirely different things is _really_ hard to parse.

File: include/sensors/core/fdc1004.hpp [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+ * We are currently only using 1 out of the 4 measurement channels.
+ *
+ * By default, CAPDAC is enabled. The max resolution of the capdac is 5 bits.
+ *
+ * Capacitance measurement is a total of 24 bits long. You must do consecutive
+ * reads of the MSB and LSB registers to get the data.
+ */
+
+namespace fdc1004_utils {
+static auto convert(uint32_t measurement, uint16_t accumulation,
```
Comment (sfoster1) [inline include/sensors/core/fdc1004.hpp:23]: `static` here is not a good idea IMO - what it will do is force the function definition to live in every single translation unit that includes this file, even transitively. These are just plain old pure functions, just declare them without static i.e. `auto convert(...) -> sq14_15;`

File: include/sensors/core/fdc1004.hpp [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+                    float current_offset) -> sq14_15;
+static auto calculate_capdac(uint16_t measurement, uint16_t accumulation,
+                             float current_offset) -> uint16_t;
+
+// Capacitance Sensor Address and Registers
+static constexpr uint16_t ADDRESS = 0x50 << 1;
```
Comment (sfoster1) [inline include/sensors/core/fdc1004.hpp:29]: [`constexpr` implies `const` implies `static`](https://stackoverflow.com/a/50489130) on variables at namespace scope, can just be `constexpr`

File: include/sensors/core/fdc1004.hpp [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+
+// Constants
+static constexpr int MAX_CAPDAC_RESOLUTION = 5;
+static constexpr int MSB_SHIFT = 16;
+static constexpr uint16_t MAX_CAPDAC_OFFSET = 0x1F << MAX_CAPDAC_RESOLUTION;
+static constexpr float MAX_MEASUREMENT = 524288.0;
+// single ended capdac offset measurement in pF
+static constexpr float CAPDAC_OFFSET = 3.125;
+
+[[maybe_unused]] static auto convert(uint32_t measurement,
```
Comment (sfoster1) [inline include/sensors/core/fdc1004.hpp:54]: why `[[maybe_unused]]`?

Comment (sfoster1) [inline include/sensors/core/fdc1004.hpp:54]: as with the above functions, this really shouldn't do `static` - it's pretty confusing in a header. If you want to keep the definition in the header, use `inline`

File: include/sensors/core/fdc1004.hpp [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+                                     uint16_t accumulation,
+                                     float current_offset) -> sq14_15 {
+    // shift the 32 bit value number down to 24 bits
+    auto average =
+        static_cast<float>(measurement >> 8) / static_cast<float>(accumulation);
+    float capacitance = average / MAX_MEASUREMENT + current_offset;
+    return convert_to_fixed_point(capacitance, 15);
+}
+
+[[maybe_unused]] static auto calculate_capdac(uint16_t measurement,
```
Comment (sfoster1) [inline include/sensors/core/fdc1004.hpp:64]: why `[[maybe_unused]]`?

Comment (sfoster1) [inline include/sensors/core/fdc1004.hpp:64]: `inline` instead of `static` please

File: include/sensors/core/tasks/capacitive_sensor_callbacks.hpp [hunk 16]
```diff
@@ -0,0 +1,128 @@
...
+                     const sensor_callbacks::MaxMessageBuffer &LSB_buffer) {
+        uint32_t msb_data = 0x0;
+        uint32_t lsb_data = 0x0;
+        const auto *MSB_iter = MSB_buffer.cbegin();
+        MSB_iter =
+            bit_utils::bytes_to_int(MSB_iter, MSB_buffer.cend(), msb_data);
+        const auto *LSB_iter = LSB_buffer.cbegin();
+        LSB_iter =
+            bit_utils::bytes_to_int(LSB_iter, LSB_buffer.cend(), lsb_data);
+        measurement += (msb_data << MSB_SHIFT | lsb_data);
```
Comment (sfoster1) [inline include/sensors/core/tasks/capacitive_sensor_callbacks.hpp:36]: why `+=`? shouldn't this just be `measurement = (...)`?

File: include/sensors/core/tasks/capacitive_sensor_callbacks.hpp [hunk 17]
```diff
@@ -0,0 +1,128 @@
...
+struct ReadCapacitanceCallback {
+  public:
+    ReadCapacitanceCallback(CanClient &can_client, float current_offset)
+        : can_client{can_client}, current_offset{current_offset} {}
+
+    void handle_data(const sensor_callbacks::MaxMessageBuffer &MSB_buffer,
+                     const sensor_callbacks::MaxMessageBuffer &LSB_buffer) {
+        uint32_t msb_data = 0x0;
+        uint32_t lsb_data = 0x0;
+        const auto *MSB_iter = MSB_buffer.cbegin();
```
Comment (sfoster1) [inline include/sensors/core/tasks/capacitive_sensor_callbacks.hpp:30]: the implied bit/byte order here sort of worries me; can we use `uint16_t` for the data elements here, and widen when we combine the two values with `MSB_SHIFT`?

File: motor-control/core/utils.cpp [hunk 18]
```diff
@@ -25,3 +25,8 @@ sq0_31 fixed_point_multiply(sq31_31 a, sq0_31 b) {
     int64_t result = a * static_cast<int64_t>(b);
     return static_cast<sq0_31>((result >> 31) & 0xFFFFFFFF);
 }
+
+float fixed_point_to_float(uint32_t data, int to_radix) {
+    const uint32_t power_of_two = 2 << to_radix;
```
Comment (sfoster1) [inline motor-control/core/utils.cpp:30]: feels like it should be a float, right?

File: include/sensors/core/fdc1004.hpp [hunk 19]
```diff
@@ -0,0 +1,78 @@
...
+                                              uint16_t accumulation,
```
Comment (sfoster1) [inline include/sensors/core/fdc1004.hpp:65]: can we call this `read_count` or something? `accumulation` doesn't really give you a lot to go on, and I think it's just the number of values that we'll be reading right

File: include/sensors/core/fdc1004.hpp [hunk 20]
```diff
@@ -0,0 +1,78 @@
...
+                                              float current_offset)
+    -> uint16_t {
+    float average =
+        static_cast<float>(measurement) / static_cast<float>(accumulation);
+    float offset = (average + current_offset) / CAPDAC_OFFSET;
```
Comment (sfoster1) [inline include/sensors/core/fdc1004.hpp:70]: This is kind of a weird calculation, can we comment what it's up to? `CAPDAC_OFFSET` above is noted as being in pF, and the average + current offset values are presumably somehow related to pF, so is the output here unitless? Bit confused.

File: include/sensors/core/fdc1004.hpp [hunk 21]
```diff
@@ -0,0 +1,78 @@
...
+// Constants
+static constexpr int MAX_CAPDAC_RESOLUTION = 5;
+static constexpr int MSB_SHIFT = 16;
+static constexpr uint16_t MAX_CAPDAC_OFFSET = 0x1F << MAX_CAPDAC_RESOLUTION;
+static constexpr float MAX_MEASUREMENT = 524288.0;
+// single ended capdac offset measurement in pF
+static constexpr float CAPDAC_OFFSET = 3.125;
+
+[[maybe_unused]] static auto convert(uint32_t measurement,
+                                     uint16_t accumulation,
```
Comment (sfoster1) [inline include/sensors/core/fdc1004.hpp:55]: let's change this name too

File: include/sensors/core/fdc1004.hpp [hunk 22]
```diff
@@ -0,0 +1,78 @@
...
+                                     float current_offset) -> sq14_15 {
+    // shift the 32 bit value number down to 24 bits
```
Comment (sfoster1) [inline include/sensors/core/fdc1004.hpp:57]: this is weird - if this is a 24 bit value, it should really be in the 24 least significant bits

**Review by sfoster1 — CHANGES_REQUESTED** [review]
Had some inline notes and change requests. The structure is there but I think there's some stuff we need to change, mostly for readability.

Aside from that, it feels a bit weird to me to have this request-response thing with the host having to read a new offset, get it from can, then send it back over can, then request data, which has the offset baked into it - so why does the host ever need to see the offset value? why not have some `update_offset` can command that, sure, maybe it returns the offset, but for informational purposes only, it updated the stored value on its own.

File: include/pipettes/core/tasks/i2c_task.hpp (hunk 10 above)
Comment (Laura-Danielle) [reply include/pipettes/core/tasks/i2c_task.hpp:52]: Put in a TODO comment for a follow-up PR.

File: include/sensors/core/tasks/capacitive_sensor_callbacks.hpp [hunk 23]
```diff
@@ -0,0 +1,128 @@
...
+struct ReadOffsetCallback {
+  public:
+    ReadOffsetCallback(CanClient &can_client, float current_offset)
+        : can_client{can_client}, current_offset{current_offset} {}
+
+    void handle_data(const sensor_callbacks::MaxMessageBuffer &buffer) {
+        uint16_t data = 0x0;
+        const auto *iter = buffer.cbegin();
+        iter = bit_utils::bytes_to_int(iter, buffer.cend(), data);
+        measurement += data;
```
[the remark this answers is not in the document]
Comment (Laura-Danielle) [reply include/sensors/core/tasks/capacitive_sensor_callbacks.hpp:82]: You might take multiple readings of read offset, or you might not. Rather than having two separate callbacks for a single read of capacitance/capdac respectively, I decided to just combine it into one.

File: include/sensors/core/fdc1004.hpp (hunk 14 above)
Comment (Laura-Danielle) [reply include/sensors/core/fdc1004.hpp:54]: Clang was getting mad at me previously saying that the callback wasn't always used -- I think changing things from using a virtual base class fixed it.

File: include/sensors/core/tasks/capacitive_sensor_callbacks.hpp [hunk 24]
```diff
@@ -0,0 +1,94 @@
+#pragma once
+#include <iostream>
```
Comment (sfoster1) [inline include/sensors/core/tasks/capacitive_sensor_callbacks.hpp:2]: use `LOG`
```suggestion
```

File: include/sensors/core/tasks/capacitive_sensor_callbacks.hpp [hunk 25]
```diff
@@ -0,0 +1,94 @@
...
+        measurement += (msb_data << MSB_SHIFT | lsb_data);
+    }
+
+    void send_to_can() {
+        uint32_t capacitance =
+            convert_capacitance(measurement, number_of_reads, current_offset);
+        auto message = can_messages::ReadFromSensorResponse{
+            {}, SensorType::capacitive, capacitance};
+        can_client.send_can_message(can_ids::NodeId::host, message);
+        std::cout << "CAPACITANCE " << static_cast<int>(capacitance) << "\n";
```
Comment (sfoster1) [inline include/sensors/core/tasks/capacitive_sensor_callbacks.hpp:46]:
```suggestion
        LOG("CAPACITANCE %d\n", static_cast<int>(capacitance));
```

File: include/sensors/core/tasks/capacitive_sensor_callbacks.hpp [hunk 26]
```diff
@@ -0,0 +1,94 @@
...
+        if (capacitance > zero_threshold || capacitance < -zero_threshold) {
+            std::cout << "greater than threshold " << "\n";
```
Comment (sfoster1) [inline include/sensors/core/tasks/capacitive_sensor_callbacks.hpp:48]:
```suggestion
            LOG("greater than threshold\n");
```

File: include/sensors/core/tasks/capacitive_sensor_callbacks.hpp [hunk 27]
```diff
@@ -0,0 +1,94 @@
...
+            auto capdac = update_capdac(capacitance, current_offset);
+            current_offset = capdac;
+            std::cout << "current offset " << static_cast<int>(current_offset)  << "\n";
```
Comment (sfoster1) [inline include/sensors/core/tasks/capacitive_sensor_callbacks.hpp:51]:
```suggestion
            LOG("current offset %d\n", static_cast<int>(current_offset));
```

File: include/sensors/core/tasks/capacitive_sensor_task.hpp [hunk 28]
```diff
@@ -0,0 +1,155 @@
...
+        /**
+         * The FDC1004 sensor has an offset register and
+         * a measurement register. The CAPDAC offset is used
+         * to account for any additional baseline noise that
+         * may occur on the capacitve sensor.
+         *
+         */
+        LOG("Received request to read from %d sensor\n", m.sensor);
+        capdac_offset = capacitance_handler.get_offset();
+        std::cout << "capdac  " << static_cast<int>(capdac_offset) << "\n";
```
Comment (sfoster1) [inline include/sensors/core/tasks/capacitive_sensor_task.hpp:63]:
```suggestion
        LOG("capdac  %d\n", static_cast<int>(capdac_offset));
```

File: include/sensors/core/tasks/capacitive_sensor_task.hpp [hunk 29]
```diff
@@ -0,0 +1,155 @@
...
+        if (bool(m.offset_reading)) {
+            std::cout << "offset reading  " << "\n";
```
Comment (sfoster1) [inline include/sensors/core/tasks/capacitive_sensor_task.hpp:65]:
```suggestion
            LOG("offset reading \n");
```

**Review by sfoster1 — CHANGES_REQUESTED** [review]
Looks great, gotta remove `#include <iostream>` and use of `cout` in a couple places though

**Review by sfoster1 — APPROVED** [review]
Awesome!
